odd_collector/adapters/elasticsearch/adapter.py

```diff
@@ -31,29 +33,137 @@ def get_data_source_oddrn(self) -> str:
         return self.__oddrn_generator.get_data_source_oddrn()
 
     def get_datasets(self) -> Iterable[DataEntity]:
+        logger.debug("Collect dataset")
         result = []
         indices = self.__get_indices()
+
+        logger.debug(f"Indeces are {indices}")
+
+        logger.debug("Process indeces")
         for index in indices:
             mapping = self.__get_mapping(index["index"])[index["index"]]
+            logger.debug(f"Mapping for index {index['index']} is {mapping}")
             try:
                 result.append(self.__process_index_data(index, mapping))
             except KeyError as e:
                 logging.warning(
                     f"Elasticsearch adapter failed to process index {index}: KeyError {e}"
                 )
+
+        logger.debug("Process data streams and their templates")
+        all_data_streams = self.__get_data_streams()
+
+        logger.debug("Build template to data stream mapping")
+        templates_info = self.get_templates_from_data_streams(all_data_streams)
+
+        for template, data_streams in templates_info.items():
+            template_meta = self.__get_data_stream_templates_info(template).get(
+                "index_templates"
+            )
+
+            data_stream_entities = []
+            logger.debug(f"Template {template} has metadata {template_meta}")
+
+            for data_stream in data_streams:
+                logger.debug(
+                    f"Data stream {data_stream['name']} has template {template}"
+                )
+
+                lifecycle_policies = self.__get_rollover_policy(data_stream)
+                stream_data_entity = map_data_stream(
+                    data_stream,
+                    template_meta,
+                    lifecycle_policies,
+                    self.__oddrn_generator,
+                )
+
+                data_stream_entities.append(stream_data_entity)
+
+            result.extend(data_stream_entities)
+
+            logger.debug(f"Create template data entity {template}")
+
+            data_streams_oddrn = [item.oddrn for item in data_stream_entities]
+            logger.debug(f"List of data streams oddrn {data_streams_oddrn}")
+
+            template_entity = map_data_stream_template(
+                template_meta, data_streams_oddrn, self.__oddrn_generator
+            )
+            result.append(template_entity)
+
         return result
 
+    def __get_rollover_policy(self, stream_data: Dict) -> Optional[Dict]:
+        backing_indices = [
+            index_info["index_name"] for index_info in stream_data["indices"]
+        ]
+        for index in backing_indices:
+
+            index_settings = self.__es_client.indices.get(index=index)
+            lifecycle_policy = index_settings[index]["settings"]["index"].get(
```

Comment:
Optionally take a look at `get_in/get_lax` function from `funcy`, i think it is suitable for cases like that and less error prone (in case of missed keys)
```
lifecycle_policy = get_lax(index_settings, [index, "settings", "index", "lifecycle"])
```